prio_processor/spark/commands.py

```diff
@@ -93,14 +91,18 @@ def generate(
     num_partitions = math.ceil(dataset_estimate_mb / partition_size_mb)
     click.echo(f"writing {num_partitions} partitions")
 
-    # caching is required for a stable id, unless we take advantage of partitioned writing
-    repartitioned = shares.repartitionByRange(num_partitions, "id").cache()
-    repartitioned.select("id", F.base64("shares.a").alias("payload")).write.json(
-        output_a, mode="overwrite"
-    )
-    repartitioned.select("id", F.base64("shares.b").alias("payload")).write.json(
-        output_b, mode="overwrite"
+    # try to be efficient without caching by repartitioning
+    repartitioned = (
+        shares.withColumn(
+            "shares",
+            F.map_from_arrays(
+                F.array(F.lit("a"), F.lit("b")), F.array("shares.a", "shares.b")
+            ),
+        )
+        .repartitionByRange(num_partitions, "id")
+        .select("id", F.explode("shares").alias("server_id", "payload"))
     )
+    repartitioned.write.partitionBy("server_id").json(output, mode="overwrite")
```

Comment:
This is significantly more efficient than the previous version since it only requires the transformation to be materialized once, and doesn't require caching an intermediate result.